File: Numerical Integration/integration.f
program compute_area
    implicit none
    real(8), parameter :: pi = 4.0_8 * atan(1.0_8) ! Define π
    real(8) :: a, b, area, exact_area, difference
    integer :: n, i
    real(8) :: h, x, sum

    ! Define the limits of integration
    a = -pi           ! -π
    b = (2.0_8 / 3.0_8) * pi  ! 2/3π

    ! Number of intervals for the trapezoidal rule
    n = 1000

    ! Step size
    h = (b - a) / n

    ! Initialize the sum for the trapezoidal rule
    sum = 0.5_8 * (sin(a) + sin(b))

    ! Compute the sum for the trapezoidal rule
    do i = 1, n-1
        x = a + i * h
        sum = sum + sin(x)
    end do

    ! Compute the area using the trapezoidal rule
    area = sum * h

    ! Print the computed area
    print *, 'The area between -π and 2/3π for sin(x) is: ', area

    ! Validation: Compute the exact area using the antiderivative of sin(x)
    ! The antiderivative of sin(x) is -cos(x)
    exact_area = (-cos(b)) - (-cos(a))

    ! Print the exact area
    print *, 'The exact area computed using the antiderivative is: ', exact_area

    ! Compare the numerical and exact results
    difference = abs(area - exact_area)
    print *, 'Difference between numerical and exact area: ', difference

end program compute_area

File: Numerical Integration/integration.java
public class ComputeArea {
    // Function to compute sin(x)
    public static double integrand(double x) {
        return Math.sin(x);
    }

    // Function to compute the area using the trapezoidal rule
    public static double trapezoidalRule(double a, double b, int n) {
        double h = (b - a) / n; // Step size
        double sum = 0.5 * (integrand(a) + integrand(b)); // First and last terms

        for (int i = 1; i < n; i++) {
            double x = a + i * h;
            sum += integrand(x);
        }

        return sum * h;
    }

    public static void main(String[] args) {
        // Define the limits of integration
        double a = -Math.PI; // -π
        double b = (2.0 / 3.0) * Math.PI; // 2/3π

        // Number of intervals for the trapezoidal rule
        int n = 1000; // Increase for better accuracy

        // Compute the area using the trapezoidal rule
        double area = trapezoidalRule(a, b, n);

        // Print the computed area
        System.out.printf("The area between -π and 2/3π for sin(x) is: %.6f%n", area);

        // Validation: Compute the exact area using the antiderivative of sin(x)
        // The antiderivative of sin(x) is -cos(x)
        double exactArea = (-Math.cos(b)) - (-Math.cos(a));

        // Print the exact area
        System.out.printf("The exact area computed using the antiderivative is: %.6f%n", exactArea);

        // Compare the numerical and exact results
        double difference = Math.abs(area - exactArea);
        System.out.printf("Difference between numerical and exact area: %.6f%n", difference);
    }
}
